mopidy_tidal/library.py

```diff
@@ -157,104 +151,136 @@ def search(self, query=None, uris=None, exact=False):
-                        pass
-
-                images[uri] = uri_images or ()
-            except AttributeError:
-                logger.error("AttributeError when processing URI %r" % uri)
-            except HTTPError as err:
-                logger.error("HTTPError when processing URI %r" % uri)
+                images[uri] = self._get_images(uri)
+            except (AssertionError, AttributeError, HTTPError) as err:
+                logger.error(
+                    "%s when processing URI %r: %s",
+                    type(err), uri, err)
 
+        self._image_cache.update(images)
         return images
 
     def lookup(self, uris=None):
         logger.info("Lookup uris %r", uris)
-        session = self.backend._session
         if isinstance(uris, str):
             uris = [uris]
         if not hasattr(uris, '__iter__'):
             uris = [uris]
 
         tracks = []
-        for uri in uris:
+        cache_updates = {}
+
+        for uri in (uris or []):
+            data = []
             try:
                 parts = uri.split(':')
-                if uri.startswith('tidal:track:'):
-                    if uri in self.track_cache:
-                        tracks.append(self.track_cache[uri])
-                    else:
-                        tracks += self._lookup_track(session, parts)
-                elif uri.startswith('tidal:album'):
-                    tracks += self._lookup_album(session, parts)
-                elif uri.startswith('tidal:artist'):
-                    tracks += self._lookup_artist(session, parts)
-                elif uri.startswith('tidal:playlist'):
-                    tracks += self._lookup_playlist(session, parts)
-
-            except AttributeError:
-                logger.error("AttributeError when processing URI %r" % uri)
+                item_type = parts[1]
+                cache_name = f'_{parts[1]}_cache'
+                cache_miss = True
+
+                try:
+                    data = getattr(self, cache_name)[uri]
+                    cache_miss = not bool(data)
+                except (AttributeError, KeyError):
+                    pass
+
+                if cache_miss:
+                    try:
+                        lookup = getattr(self, f'_lookup_{parts[1]}')
+                    except AttributeError:
+                        continue
+
+                    data = cache_data = lookup(self._session, parts)
+                    cache_updates[cache_name] = cache_updates.get(cache_name, {})
+                    if item_type == 'playlist':
+                        # Playlists should be persisted on the cache as objects,
+                        # not as lists of tracks. Therefore, _lookup_playlist
+                        # returns a tuple that we need to unpack
+                        data, cache_data = data
+
+                    cache_updates[cache_name][uri] = cache_data
+
+                tracks += data if hasattr(data, '__iter__') else [data]
```

Comment:
We should change L251 to the following:
```
if item_type == 'playlist':
    tracks += data.tracks
else:
    tracks += data if hasattr(data, '__iter__') else [data]
```
or make sure tuple is unpacked when no cache miss occurs. Otherwise, we will still get the error:
```
ERROR    [Core-7] mopidy.core.library TidalBackend backend returned bad data: Expected a list of Track, not [Playlist
```